### DISCUSSION TITLE
How can I use inference.crop() with a list of segments for SpeechBrain/NeMo models?

### DISCUSSION THREAD

**Discussion by RakeshReddyKondeti** [pyannote/pyannote-audio#1956] [Q&A]
Hello,

I am using the following approach (from the official examples) to extract embeddings for specific segments using the SpeechBrain ECAPA model:

```python
import torch
from pyannote.audio.pipelines.speaker_verification import PretrainedSpeakerEmbedding
model = PretrainedSpeakerEmbedding(
    "speechbrain/spkrec-ecapa-voxceleb",
    device=torch.device("cuda"))

from pyannote.audio import Audio
from pyannote.core import Segment
audio = Audio(sample_rate=16000, mono="downmix")

# extract embedding for a speaker speaking between t=3s and t=6s
speaker1 = Segment(3., 6.)
waveform1, sample_rate = audio.crop("audio.wav", speaker1)
embedding1 = model(waveform1[None])
```

With `pyannote` models, I can use the [`Inference`](https://github.com/pyannote/pyannote-audio/blob/ad15cf41dde06f833646b9b73091887d60c82576/src/pyannote/audio/core/inference.py#L46-L667) wrapper to easily extract embeddings for segments, like this:

```python
from pyannote.audio import Inference
from pyannote.core import Segment

inference = Inference("pyannote/embedding")
embedding = inference.crop("audio.wav", Segment(3, 6))
# or for multiple segments:
embeddings = inference.crop("audio.wav", [Segment(1, 4), Segment(10, 15)])
```

However, the [`Inference`](https://github.com/pyannote/pyannote-audio/blob/ad15cf41dde06f833646b9b73091887d60c82576/src/pyannote/audio/core/inference.py#L46-L667) wrapper expects a model that is a subclass of [`pyannote.audio.core.model.Model`](https://github.com/pyannote/pyannote-audio/blob/ad15cf41dde06f833646b9b73091887d60c82576/src/pyannote/audio/core/model.py#L69-L651), and does not work with SpeechBrain or NeMo models (like ECAPA or Titanet), which are wrapped by [`PretrainedSpeakerEmbedding`](https://github.com/pyannote/pyannote-audio/blob/ad15cf41dde06f833646b9b73091887d60c82576/src/pyannote/audio/pipelines/speaker_verification.py#L719).

### Question:
Is there a recommended way to pass a list of segments at once (like with [inference.crop](https://github.com/pyannote/pyannote-audio/blob/ad15cf41dde06f833646b9b73091887d60c82576/src/pyannote/audio/core/inference.py#L417)) and efficiently extract embeddings for all of them when using SpeechBrain or NeMo embedding models? 

Thank you!

**Comment by hbredin** [comment]
Unfortunately, we are deprecating support for Speechbrain and NeMo models for this very reason: they are not pyannote-native models and cannot easily be used within this Inference class...

**Comment by arcticloud** [comment]
Right — `Inference` only wraps pyannote-native models, so for a SpeechBrain/NeMo `PretrainedSpeakerEmbedding` you don't get `.crop([seg, seg, ...])`. Just crop each segment yourself and **batch them into one forward pass**:

```python
import torch
from pyannote.audio import Audio
from pyannote.audio.pipelines.speaker_verification import PretrainedSpeakerEmbedding
from pyannote.core import Segment

model = PretrainedSpeakerEmbedding("speechbrain/spkrec-ecapa-voxceleb",
                                   device=torch.device("cuda"))
audio = Audio(sample_rate=16000, mono="downmix")

segments = [Segment(3, 6), Segment(10, 14), Segment(20, 23)]

# crop -> pad to equal length -> stack to (N, 1, samples) -> single call
waveforms = [audio.crop("audio.wav", s)[0] for s in segments]   # each (1, samples)
maxlen = max(w.shape[-1] for w in waveforms)
batch = torch.stack([torch.nn.functional.pad(w, (0, maxlen - w.shape[-1]))
                     for w in waveforms])                        # (N, 1, samples)
embeddings = model(batch)                                        # (N, dim)
```

`embeddings[i]` is the vector for `segments[i]`. Two caveats: ECAPA pools over time so zero-padding short clips is fine, but if your segments vary wildly in length, bucket them by duration into a few batches so padding doesn't dominate. And since (as noted here) SpeechBrain/NeMo support is being deprecated in this class, if you can switch to `pyannote/embedding` you get the `Inference` convenience (including list-of-segments) back for free.
